### PR TITLE
Setup clang-tidy (C++ linter)

### PR DESCRIPTION
**Pull request by Jarred-Sumner**
### What does this PR do?

This adds clang-tidy which is a very slow C++ linter. It is slow because it first compiles all the C++ code, and then lints it.

### How did you verify your code works?

Lets see what CI says

### CODE REVIEW

**Comment by alex-demchenko** [comment]
Hi! Also PVS-Studio (static code analyzer) is pretty good. For example its used in Unreal Engine.

File: .github/workflows/bun-release.yml [hunk 1]
```diff
@@ -62,6 +62,8 @@ jobs:
         with:
           bun-version: latest
       - name: Install Dependencies
+        run: apt update && apt install -y llvm-16-dev libclang-16-dev clang-16 libssl-dev libxml2-dev build-essential ninja-build cmake
```
Comment (Electroid) [inline .github/workflows/bun-release.yml]: Edited wrong file?

**Comment by paperclover** [comment]
why are you passing NO_CODEGEN?

this build:lint script doesnt run locally and it doesnt even work in ci

**Comment by paperclover** [comment]
I have pushed a commit to make this actually run the build, but there are failing lints. some are very suspicious.

File: src/bun.js/bindings/BunPlugin.cpp [hunk 2]
```diff
@@ -309,20 +308,14 @@ extern "C" JSC::EncodedJSValue setupBunPlugin(JSC::JSGlobalObject* globalObject,
...
-            } else {
+            if (!(targetString == "node"_s || targetString == "bun"_s || targetString == "browser"_s)) {
                 JSC::throwTypeError(globalObject, throwScope, "plugin target must be one of 'node', 'bun' or 'browser'"_s);
                 return JSValue::encode(jsUndefined());
             }
         }
     }
 
-    JSFunction* setupFunction = jsCast<JSFunction*>(setupFunctionValue);
+    // JSFunction* setupFunction = jsCast<JSFunction*>(setupFunctionValue);
```
Comment (jdalton) [inline src/bun.js/bindings/BunPlugin.cpp]: Debug code ☝️ ?

File: src/bun.js/bindings/BunProcess.cpp [hunk 3]
```diff
@@ -1649,12 +1648,10 @@ static JSValue constructStdin(VM& vm, JSObject* processObject)
 {
     auto* globalObject = Bun__getDefaultGlobal();
     auto scope = DECLARE_THROW_SCOPE(vm);
-    auto* thisObject = reinterpret_cast<Zig::GlobalObject*>(globalObject);
     JSC::JSFunction* getStdioWriteStream = JSC::JSFunction::create(vm, processObjectInternalsGetStdinStreamCodeGenerator(vm), globalObject);
     JSC::MarkedArgumentBuffer args;
     args.append(JSC::jsNumber(STDIN_FILENO));
 
-    auto clientData = WebCore::clientData(vm);
```
Comment (jdalton) [inline src/bun.js/bindings/BunProcess.cpp]: Yes removal of unused vars :DDD

**Comment by jdalton** [comment]
💪 I dig this

**Review by paperclover — CHANGES_REQUESTED** [review]
putting a request changes to get this out of my inbox

issues with this as i remember a month ago are mainly that it does not solve all exposed warnings (will just be an annoyance to CI that is never paid attention to if it is always red)

**Pull request merged**